astropy/modeling/core.py

```diff
@@ -1487,23 +1506,30 @@ def prepare_inputs(self, *inputs, model_set_axis=None, equivalencies=None,
                     # input values without conversion, otherwise we raise an
                     # exception.
 
-                    if (not self.input_units_allow_dimensionless and
+                    if (not input_units_allow_dimensionless[input_name] and
                        input_unit is not dimensionless_unscaled and input_unit is not None):
                         if np.any(inputs[i] != 0):
                             raise UnitsError("Units of input '{0}', (dimensionless), could not be "
                                              "converted to required input units of "
                                              "{1} ({2})".format(self.inputs[i], input_unit,
                                                                 input_unit.physical_type))
 
```

Comment:
When evaluating a compound model it's unclear which model the message refers to. It'd be good to see the name of the model. For example:

```
g = Gaussian1D(10 * u.Jy, 2 * u.m, 1 * u.m, name='Gauss') 
aff = AffineTransformation2D([[1, 0] , [0, 1]] * u.m, name="rotation_model") 
(g&g | aff)(1*u.m, 2*u.m)
...
UnitsError: rotation_model: Units of input 'x', Jy (spectral flux density), could not be converted to 
required input units of m (length)
```
Could you add the name of the model to the error messages? 
```
raise UnitsError("{0}: Units of input '{1}', {2} ({3}), could not be "
                          "converted to required input units of "
                          "{4} ({5})".format(self.name, self.inputs[i],
                                                     inputs[i].unit,
                                                      inputs[i].unit.physical_type,
                                                      input_unit,
                                                     input_unit.physical_type))
```

Reply:
@Cadair  I accidentally hit "single comment' - in case you want to wait for the rest.

Reply:
`Model.name` is not always defined? So what do we want to do if it's `None`?